src/java/org/apache/cassandra/service/StorageProxy.java

```diff
@@ -1061,6 +1071,12 @@ public static void mutate(List<? extends IMutation> mutations,
 
         QueryInfoTracker.WriteTracker writeTracker = queryTracker().onWrite(state, false, mutations, consistencyLevel);
 
+        // Request sensors are utilized to track usages from replicas serving a write request
+        RequestSensors sensors = CassandraRelevantProperties.PROPAGATE_REQUEST_SENSORS_VIA_NATIVE_PROTOCOL.getBoolean() ?
```

Comment:
Shouldn't this use `RequestSensorsFactory` rather than creating things manually? Likewise for reads below.

Reply:
Also using `RequestSensorsFactory ` would effectively mean (at least in CNDB case) running the logic `RequestSensorsFactory#create(keyspace)` twice per request - once on the coordinator and another on the writer.

Reply:
> Unlike on writers, here what decides if we should track sensors is the REQUEST_SENSORS_VIA_NATIVE_PROTOCOL flag & whether or not sensors values exit in writer response.

I'm not sure I understand this correctly, but I'd say at this level we should always use `RequestSensorsFactory` as we're deciding how to _track_ sensors, then we should use `REQUEST_SENSORS_VIA_NATIVE_PROTOCOL` when deciding if returning them on the native protocol response. Makes sense?

> a write request at the storage proxy level could contain mutations that belong to different keyspaces. Because we have one RequestSensors per thread, I wasn't sure how to make it per ks

Good point. This is kind of an anomaly by the API point of view. The instinct says "it's an anomaly, so ignore it and assumes all keyspaces are enabled (or not)", but let me think about that.

Reply:
> I'm not sure I understand this correctly, but I'd say at this level we should always use RequestSensorsFactory as we're deciding how to track sensors, then we should use REQUEST_SENSORS_VIA_NATIVE_PROTOCOL when deciding if returning them on the native protocol response. Makes sense?

I see what you mean. What I'm trying to say is the decision criteria to track sensors on the writer is `RequestSensorsFactory ` but here it is `REQUEST_SENSORS_VIA_NATIVE_PROTOCOL`. If we disable the latter, there is no point to track sensors on coordinator (at least as far as this patch is concerned). Also, because we have a single `RequestSensorsFactory` the track/don't track decision will always be the same on writer and coordinators and it sounded redundant to rerun the logic on both components.  

Having said that, I don't have a strong opinion here, and the bigger issue is the decision `RequestSensorsFactory` per ks discussion point so I'll put this on hold until we conclude on the other one.

Reply:
actually I'm convinced - lets figure out the multiple keyspace things and I'll use `RequestSensorsFactory` across the board.

Reply:
Fixed: https://github.com/datastax/cassandra/pull/1295/commits/f9c7af504749fdea527f5e78aaa606d1c5bd55ec